packages/dtslint/src/index.ts

```diff
@@ -142,10 +157,77 @@ async function runTests(
+    );
+    if (pkg) {
+      const { packageName, packageVersion, tarballUrl } = pkg;
+      if (packageJson.nonNpm === true) {
+        (errors ??= []).push(
+          `Package ${packageJson.name} is marked as non-npm, but ${packageName} exists on npm. ` +
+            `If these types are being added to DefinitelyTyped for the first time, please choose ` +
+            `a different name that does not conflict with an existing npm package.`,
+        );
+      } else if (!packageJson.nonNpm) {
+        if (!satisfies(packageVersion, typesPackageVersion)) {
+          hasNpmVersionMismatch = true;
+          const isError = !npmVersionExemptions.has(packageDirectoryNameWithVersion);
+          const container = isError ? (errors ??= []) : (warnings ??= []);
+          container.push(
+            (isError
+              ? ""
+              : `Ignoring npm version error because ${packageDirectoryNameWithVersion} was failing when the check was added. ` +
+                `If you are making changes to this package, please fix this error:\n> `) +
+              `Cannot find a version of ${packageName} on npm that matches the types version ${typesPackageVersion}. ` +
+              `The closest match found was ${packageName}@${packageVersion}. ` +
+              `If these types are for the existing npm package ${packageName}, change the ${packageDirectoryNameWithVersion}/package.json ` +
+              `major and minor version to match an existing version of the npm package. If these types are unrelated to ` +
+              `the npm package ${packageName}, add \`"nonNpm": true\` to the package.json and choose a different name ` +
+              `that does not conflict with an existing npm package.`,
+          );
+        } else {
+          const tarballPath = joinPaths(tmpDir, `${packageName.replace(/\//g, "-")}-${packageVersion}.tgz`);
+          const unpackedPath = joinPaths(tmpDir, `${packageName}-${packageVersion}`);
+          await pipeline((await fetch(tarballUrl)).body!, fs.createWriteStream(tarballPath));
+          await fs.promises.mkdir(unpackedPath, { recursive: true });
+          await pipeline(fs.createReadStream(tarballPath), createGunzip(), tar.extract({ cwd: unpackedPath }));
+          const containerDir = (await fs.promises.readdir(unpackedPath))[0];
+          implementationPackage = {
+            packageName,
+            packageVersion,
+            tarballPath,
+            unpackedPath: joinPaths(unpackedPath, containerDir),
+          };
+        }
+      }
+    } else {
+      if (packageJson.nonNpm === "conflict") {
+        (errors ??= []).push(
+          `Package ${packageJson.name} is marked as \`"nonNpm": "conflict"\`, but no conflicting package name was ` +
+            `found on npm. These non-npm types can be makred as \`"nonNpm": true\` instead.`,
+        );
+      } else if (packageJson.nonNpm === false) {
```

Comment:
I think this should also check undefined; it's more likely that nonNpm isn't present, right? I don't think we force nonNpm to be set.

Reply:
Hm, I guess this is already a parsed header, so it has already been normalized?

Reply:
Yeah, not sure why I wrote it this way though. Likely a refactor from a switch/case or something